### ISSUE TITLE
Changing default values for options

### ISSUE THREAD

**Issue by antonleykin** [Macaulay2/M2#2240]
It would be nice to have a facility to change the default value of an optional argument. E.g., there could be a method `setDefault` with `setDefault(SVD,DivideConquer=>true)` resulting in resetting the default from `false` to `true`.

This would be useful in a scenario when a user 

- writes code that relies on popular methods (e.g., `SVD`, `gb`, `res`, etc.) without explicitly referring to strategies, and
- experiments later by tweaking strategies.

**Comment by DanGrayson** [comment]
I think that would cause problems with packages you use that call the function (unbeknownst to you) and that depend on the default values of the optional arguments being as advertised in the documentation.  I recommend not pursuing this.

**Comment by antonleykin** [comment]
Of course, what you have in mind is a possible scenario. Another common scenario is that there are 

- method `A` with an option 
  - whose value doesn't affect the output at all, or
  - doesn't affect it conceptually (e.g., in numerical linear algebra approximations are produced) 
- method `B` in a package that uses `A` and doesn't have all the "tuning" parameters (that you may want to pass to its subroutines like `A`) as optional arguments --- having all these would imply gobs of cumbersome unreadable code.

Now, writing a piece of code using `B`, 
  - one has no freedom to experiment with different settings for options in the subroutines like `A`  
  - reengineering the package of `B` for this purpose is a questionable proposition

**Comment by DanGrayson** [comment]
We already have a provision for packages to give access to the values of tuning parameters to users: configuration options, set up by the Configuration option to newPackage.  That's preferable to allowing one package to change things about another package, which will lead to confusion when many packages are changing many tunable parameters of many packages.

**Comment by mahrud** [comment]
I've wanted something like what Anton suggests for ages. A possibility is to "replace" the method, for instance in the same way that [profile](https://faculty.math.illinois.edu/Macaulay2/doc/Macaulay2/share/doc/Macaulay2/Macaulay2Doc/html/_profile.html) does, with a new function that has different default values for options. I think there may be a solution involving symbols and frames in Macaulay2 to address Dan's concerns, but that mechanism still isn't documented anywhere as far as I know.

**Comment by antonleykin** [comment]
That's right, @mahrud --- I thought about this workaround as well. The immediate problem is in this not affecting the method that is used in a package: e.g., I can't `profile` a subroutine used in a package.
```
i1 : oldSVD := SVD

o1 = SVD

o1 : MethodFunctionWithOptions

i2 : SVD := profile oldSVD

o2 = -*Function[/home/aleykin3/M2repos/anton-LAcore/M2/Macaulay2/m2/profile.m2:13:11-17:15]*-

o2 : FunctionClosure

i3 : needsPackage "NumericalLinearAlgebra"

o3 = NumericalLinearAlgebra

o3 : Package

i4 : numericalRank random(RR^2,RR^3)

o4 = 2
```

**Comment by antonleykin** [comment]
@DanGrayson, I agree that resetting default options has more of a "configuration" or "environment" variable flavor. Yet specifying a `Configuration` option in a package would not address, say, a desire to change `Strategy` in all `gb` subroutines involved everywhere in a user's script.

What one might want is some way of specifying a "configuration" for a particular method, be it in a package or `Core`.

**Comment by DanGrayson** [comment]
Well, we could probably arrange for the Core package to have configuration options, too.

**Comment by mahrud** [comment]
It's not feasible to change _every_ option in the Core package to be a configuration option.

> That's right, @mahrud --- I thought about this workaround as well. The immediate problem is in this not affecting the method that is used in a package: e.g., I can't `profile` a subroutine used in a package.

I think this also goes back to frames and symbols; if we could rewrite a symbol defined in a different package (= in a different frame?) then this would be doable. I think this would allow even more fine tuning, for instance you can have `numericalRank` and `numericalRank2`, each using a different version of `SVD` side by side without the need to change some configuration option. Again, would be great if frames and symbols were more malleable.

**Comment by DanGrayson** [comment]
> It's not feasible to change _every_ option in the Core package to be a configuration option.
> 

Yes, but we wouldn't want to do that.  For example, none of these should be configurable:
```m2

i1 : options newPackage

o1 = OptionTable{Authors => {}                               }
                 AuxiliaryFiles => false
                 CacheExampleOutput => null
                 Certification => null
                 Configuration => {}
                 Date => null
                 DebuggingMode => false
                 Headline => null
                 HomePage => null
                 InfoDirSection => Macaulay2 and its packages
                 Keywords => {Uncategorized}
                 OptionalComponentsPresent => null
                 PackageExports => {}
                 PackageImports => {}
                 Reload => false
                 UseCachedExampleOutput => null
                 Version => 0.0

o1 : OptionTable
```

**Comment by DanGrayson** [comment]
Here's how to make SVD of a matrix always use true for that option:

```m2
i1 : g := lookup(SVD,Matrix);

i2 : SVD Matrix := o -> f -> (g new OptionTable from { DivideConquer => true }) f;
```

**Comment by DanGrayson** [comment]
Look at the optional arguments for `gb`:
```m2

i19 : options gb

o19 = OptionTable{Algorithm => Inhomogeneous        }
                  BasisElementLimit => infinity
                  ChangeMatrix => false
                  CodimensionLimit => infinity
                  DegreeLimit => {}
                  GBDegrees => null
                  HardDegreeLimit => null
                  Hilbert => null
                  MaxReductionCount => 10
                  PairLimit => infinity
                  StopBeforeComputation => false
                  StopWithMinimalGenerators => false
                  Strategy => {}
                  SubringLimit => infinity
                  Syzygies => false
                  SyzygyLimit => infinity
                  SyzygyRows => infinity

o19 : OptionTable
```

Only one of them can be changed without affecting the mathematical value of the answer and breaking code all over the place: "Strategy".  So one way to allow it to be changed is to add a global variable called `gbStrategy` that the user can set.

**Comment by mahrud** [comment]
I think going over the nearly 100 optional argument names (many of them appearing in multiple methods) and deciding which should be global configuration is a waste of time and will add many new symbols. Do you really want to add nearly 50 global configs like `gbDefaultStrategy`, `syzDefaultStrategy`, `intersectDefaultStrategy`, etc etc? This is going to be a pain to document and just seems like a very bad idea.

I like the way you showed for replacing the default for (SVD, Matrix). A similar simple method for a whole method is what is desired.

> Only one of them can be changed without affecting the mathematical value of the answer and breaking code all over the place

I disagree. Multiple other options might be useful to change. I think your point of view is limited by what you consider useful, as opposed to giving users the freedom to do things they find useful. For instance, what if I wanted to _add a new option_ for `gb`? I think the simplest and most general solution is providing a way to replace a method, either within a specific frame or globally.

**Comment by mahrud** [comment]
> Here's how to make SVD of a matrix always use true for that option:
> 
> ```m2
> i1 : g := lookup(SVD,Matrix);
> 
> i2 : SVD Matrix := o -> f -> (g new OptionTable from { DivideConquer => true }) f;
> ```

Oh, there's a big problem with this that I didn't realize earlier: Anton's point was to change the _default_ value, not _force_ change it. Your solution will always override that option, not just change the default. This is not sufficient, at least for my needs.

**Comment by DanGrayson** [comment]
That can be done this way:
```m2

i1 : options SVD

o1 = OptionTable{DivideConquer => false}

o1 : OptionTable

i2 : newopts := options SVD ++ { DivideConquer => true}

o2 = OptionTable{DivideConquer => true}

o2 : OptionTable

i3 : g := lookup(SVD,Matrix);

i4 : SVD Matrix := o -> m -> (g (newopts ++ o)) m;
```

**Comment by DanGrayson** [comment]
Yes, by the time the function on line `i2` gets called, it is no longer known which optional argument values are there because they are the default, and which are there because the user gave the value explicitly.

**Comment by antonleykin** [comment]
Thanks @DanGrayson and @mahrud !
Both ways of setting the default options work. Also, now I have more knowledge about `lookup` and `frame`!
We may want to share this during M2internals and have a discussion on whether a convenience method for setting defaults should be provided to a user (perhaps with some "warnings" issued).
